test/unit/spec/nbextensions/appCell2/widgets/appCellWidget-spec.js

```diff
@@ -1,152 +1,662 @@
+                        });
+                        expect(busEmissions).toContain([
+                            jcm.MESSAGE_TYPE.START_UPDATE,
+                            { [jcm.PARAM.JOB_ID]: TEST_JOB },
+                        ]);
+                    });
+                });
+
+                it('responds to launch errors', function () {
+                    const runStatusArgs = { event: 'error', cell_id: this.cell_id };
+
+                    return new Promise((resolve) => {
+                        spyOn(Narrative, 'saveNotebook').and.callFake(() => {
+                            resolve();
+                        });
+                        send_RUN_STATUS({ bus: this.bus, runStatusArgs });
+                    }).then(() => {
+                        // FSM state to { mode: 'error', stage: 'launching' }
+                        expect(this.appCellWidgetInstance.__fsm().getCurrentState().state).toEqual({
+                            mode: 'error',
+                            stage: 'launching',
+                        });
+                        const exec = this.appCellWidgetInstance.model.getItem('exec');
+                        expect(exec.launchState).toEqual({
+                            ...runStatusArgs,
+                            event_at: 1234567890,
+                        });
+                        // action button should be reRunApp
+                        expect(
+                            this.kbaseNode.querySelector(`.kb-rcp__action-button.-rerun`)
+                        ).not.toHaveClass('hidden');
+                        // expect the error tab to be visible
+                        expect(
+                            this.kbaseNode.querySelector(`.kb-rcp__tab-button[data-button="error"]`)
+                        ).not.toHaveClass('hidden');
+                    });
                 });
+
+                // TODO: ATM the app cell doesn't respond to invalid messages
+                xit('responds to crazy messages', function () {
+                    const runStatusArgs = { event: 'THE APOCALYPSE', cell_id: this.cell_id };
+                    send_RUN_STATUS({
+                        bus: this.bus,
+                        runStatusArgs,
+                    });
+                    // message doesn't get saved and there's no narrative save, so how to know when
+                    // message is processed?
+                    expect(this.appCellWidgetInstance.__fsm().getCurrentState().state).toEqual({
+                        mode: 'editing',
+                        params: 'complete',
+                        code: 'built',
+                    });
+                });
+            });
+
+            describe('job status updates', () => {
+                // ensure that the app cell goes from { mode: 'processing', stage: 'launched', }
+                // to the appropriate state on receiving a job status message
+                Object.keys(JobsData.jobsById).forEach((jobId) => {
```

Comment:
Tests to ensure that the app cell can receive a job status message and update itself to the correct state. This turned up another FSM issue. 😖